File: adapters/go/i18nt.go
package i18nt

import (
	"encoding/json"
	"fmt"
	"os"
	"regexp"
	"strings"
)

type I18nt struct {
	Translations map[string]interface{}
	Locale       string
}

func NewI18nt(filePath string) (*I18nt, error) {
	file, err := os.ReadFile(filePath)
	if err != nil {
		return nil, err
	}

	var data struct {
		Language     string                 `json:"language"`
		Translations map[string]interface{} `json:"translations"`
	}

	if err := json.Unmarshal(file, &data); err != nil {
		return nil, err
	}

	return &I18nt{
		Translations: data.Translations,
		Locale:       data.Language,
	}, nil
}

func (i *I18nt) T(path string, params map[string]interface{}) string {
	keys := strings.Split(path, ".")
	var val interface{} = i.Translations

	for _, k := range keys {
		if m, ok := val.(map[string]interface{}); ok {
			val = m[k]
		} else {
			return path
		}
	}

	strVal, ok := val.(string)
	if !ok {
		return path
	}

	return i.formatICU(strVal, params)
}

func (i *I18nt) formatICU(template string, params map[string]interface{}) string {
	result := template

	// 1. Simple Variables: {name}
	for k, v := range params {
		result = strings.ReplaceAll(result, "{"+k+"}", fmt.Sprintf("%v", v))
	}

	// 2. Basic Plural: {count, plural, =0{...} other{...}}
	pluralRegex := regexp.MustCompile(`\{(\w+),\s*plural,\s*(.*?)\}`)
	result = pluralRegex.ReplaceAllStringFunc(result, func(m string) string {
		match := pluralRegex.FindStringSubmatch(m)
		varName := match[1]
		optionsStr := match[2]
		
		count := 0
		if c, ok := params[varName].(int); ok {
			count = c
		} else if c, ok := params[varName].(float64); ok {
			count = int(c)
		}

		options := parseOptions(optionsStr)
		exactKey := fmt.Sprintf("=%d", count)
		
		res, ok := options[exactKey]
		if !ok {
			res = options["other"]
		}

		return strings.ReplaceAll(res, "#", fmt.Sprintf("%d", count))
	})

	// 3. Select: {gender, select, male{...} female{...} other{...}}
	selectRegex := regexp.MustCompile(`\{(\w+),\s*select,\s*(.*?)\}`)
	result = selectRegex.ReplaceAllStringFunc(result, func(m string) string {
		match := selectRegex.FindStringSubmatch(m)
		varName := match[1]
		optionsStr := match[2]
		
		val := fmt.Sprintf("%v", params[varName])
		options := parseOptions(optionsStr)
		
		res, ok := options[val]
		if !ok {
			res = options["other"]
		}
		return res
	})

	return result
}

func parseOptions(s string) map[string]string {
	options := make(map[string]string)
	optRegex := regexp.MustCompile(`(=?\w+)\s*\{(.*?)\}`)
	matches := optRegex.FindAllStringSubmatch(s, -1)
	for _, m := range matches {
		options[m[1]] = m[2]
	}
	return options
}

File: adapters/dart/i18nt.dart
import 'dart:convert';
import 'package:flutter/services.dart';

class I18nt {
  final Map<String, dynamic> _data;

  I18nt(this._data);

  /// 静态加载方法，支持从 assets 加载 JSON
  static Future<I18nt> load(String path) async {
    final String content = await rootBundle.loadString(path);
    final data = json.decode(content);
    return I18nt(data['translations'] ?? {});
  }

  /// 使用点语法路径获取翻译
  String t(String path, [Map<String, dynamic>? params]) {
    List<String> parts = path.split('.');
    dynamic current = _data;

    for (var part in parts) {
      if (current is Map && current.containsKey(part)) {
        current = current[part];
      } else {
        return '[$path]';
      }
    }

    if (current is String) {
      String result = current;
      if (params != null) {
        params.forEach((key, value) {
          result = result.replaceAll('{$key}', value.toString());
        });
      }
      return result;
    }

    return current?.toString() ?? '[$path]';
  }

  /// 支持方括号语法：t['common']['save']
  dynamic operator [](String key) {
    var val = _data[key];
    if (val is Map<String, dynamic>) {
      return I18nt(val);
    }
    return val;
  }

  @override
  String toString() => _data.toString();
}
